tests/test_synced_collection.py

```diff
@@ -0,0 +1,616 @@
+        sd[key] = d
+        assert len(sd) == 1
+        assert sd[key] == d
+        d1 = sd.pop(key)
+        assert len(sd) == 0
+        assert d == d1
+        with pytest.raises(KeyError):
+            sd[key]
+        d2 = sd.pop(key, 'default')
+        assert len(sd) == 0
+        assert d2 == 'default'
+
+    def test_popitem(self, synced_collection, testdata):
+        sd = synced_collection
+        key = 'pop'
+        d = testdata
+        sd[key] = d
+        assert len(sd) == 1
+        assert sd[key] == d
+        key1, d1 = sd.popitem()
+        assert len(sd) == 0
+        assert key == key1
+        assert d == d1
+        with pytest.raises(KeyError):
+            sd[key]
+
+    def test_values(self, synced_collection, testdata):
+        sd = synced_collection
+        data = {'value1': testdata, 'value_nested': {'value2': testdata}}
+        sd.reset(data)
+        assert 'value1' in sd
+        assert 'value_nested' in sd
+        for val in sd.values():
+            assert not isinstance(val, self._type)
+            assert val in data.values()
+
+    def test_items(self, synced_collection, testdata):
+        sd = synced_collection
+        data = {'item1': testdata, 'item_nested': {'item2': testdata}}
+        sd.reset(data)
+        assert 'item1' in sd
+        assert 'item_nested' in sd
+        for key, val in sd.items():
+            assert sd[key] == data[key]
+            assert not isinstance(val, self._type)
+            assert (key, val) in data.items()
+
+    def test_reset(self, synced_collection, testdata):
+        sd = synced_collection
+        key = 'reset'
+        d = testdata
+        sd[key] = d
+        assert len(sd) == 1
+        assert sd[key] == d
+        sd.reset()
+        assert len(sd) == 0
+        d1 = testdata
+        sd.reset({'reset': d1})
+        assert len(sd) == 1
+        assert sd[key] == d1
+
+        # invalid input
+        with pytest.raises(ValueError):
+            sd.reset([0, 1])
+
+    def test_attr_dict(self, synced_collection, testdata):
+        sd = synced_collection
+        key = 'test'
+        d = testdata
+        sd[key] = d
+        assert len(sd) == 1
+        assert key in sd
+        assert sd[key] == d
+        assert sd.get(key) == d
+        assert sd.test == d
+        del sd.test
+        assert len(sd) == 0
+        assert key not in sd
+        d1 = testdata
+        key = 'test2'
+        sd.test2 = d1
+        assert len(sd) == 1
+        assert key in sd
+        assert sd[key] == d1
+        assert sd.get(key) == d1
+        assert sd.test2 == d1
+        with pytest.raises(AttributeError):
+            sd.not_exist
+
+        # deleting a protected attribute
+        sd.load()
+        del sd._parent
+        # deleting _parent will lead to recursion as _parent is treated as key
+        # load() will check for _parent and __getattr__ will call __getitem__ which calls load()
+        with pytest.raises(RecursionError):
+            sd.load()
```

Comment:
I am not sure about this test (deleting a protected attribute). Please suggest any other way to test this.